### PR TITLE
Justin Wu submission

### PR DESCRIPTION
**Pull request by notJustinWu**
For this run, I implemented QK-norm, where we apply RMS norm to the Q and K before running attention. I also implemented weight-tying. Finally, I added a standard Mixture of Experts (MoE) design with 4 experts per block, with each pass taking weights from the top 2 experts. I also did some hyper-parameter tuning on the learning rate, and changed the angle for RoPE embeddings to 500000.

### CODE REVIEW

**Comment by aggwall** [comment]
excellent job!! i'm super impressed by how you implemented the sophisticated Mixture of Experts architectural optimization.

**Comment by notJustinWu** [comment]
> excellent job!! i'm super impressed by how you implemented the sophisticated Mixture of Experts architectural optimization.

Thanks Mr. Agrawal, I appreciate your support.
